integration-tests/src/test/java/com/datastax/oss/driver/mapper/EntityPolymorphismIT.java

```diff
@@ -0,0 +1,972 @@
+  }
+
+  // define parent interface with dao methods.
+  @DefaultNullSavingStrategy(NullSavingStrategy.SET_TO_NULL)
+  interface BaseDao<T> {
+    @Insert
+    void save(T t);
+
+    @Select
+    T findById(UUID id);
+
+    @Delete
+    void delete(T t);
+
+    @SetEntity
+    void bind(T t, BoundStatementBuilder builder);
+
+    @GetEntity
+    T one(ResultSet result);
+
+    @Update
+    void update(T t);
+  }
+
+  // another parent interface with dao methods
+  interface WriteTimeDao<Y extends WriteTimeProvider> extends BaseDao<Y> {
+    @Insert(timestamp = ":writeTime")
+    void saveWithTime(Y y, long writeTime);
+  };
+
+  @Dao
+  interface RectangleDao extends BaseDao<Rectangle> {};
+
+  // Define an intermediate interface with same type variable name to ensure
+  // this doesn't cause any issue in code generation.
+  interface ArbitraryInterface<Y extends Number> {
+    default long increment(Y input) {
+      return input.longValue() + 1;
+    }
+  }
+
+  @Dao
+  interface SquareDao extends WriteTimeDao<Square>, ArbitraryInterface<Float> {};
+
+  @Dao
+  interface CircleDao extends WriteTimeDao<Circle> {}
+
+  @Dao
+  interface SphereDao extends WriteTimeDao<Sphere> {};
+
+  interface NamedDeviceDao<T extends Device> extends BaseDao<T> {
+    @Query("UPDATE ${qualifiedTableId} SET name = :name WHERE device_id = :id")
+    void updateName(String name, UUID id);
+
+    @Query("SELECT * FROM ${qualifiedTableId} WHERE device_id = :id")
+    CompletableFuture<T> findByIdQueryAsync(UUID id);
+  }
+
+  @Dao
+  interface DeviceDao extends NamedDeviceDao<Device> {}
+
+  @Dao
+  interface TrackedDeviceDao extends NamedDeviceDao<TrackedDevice> {}
+
+  @Dao
+  interface SimpleDeviceDao extends BaseDao<SimpleDevice> {}
+
+  @Mapper
+  public interface TestMapper {
+    @DaoFactory
+    CircleDao circleDao(@DaoKeyspace CqlIdentifier keyspace);
+
+    @DaoFactory
+    RectangleDao rectangleDao(@DaoKeyspace CqlIdentifier keyspace);
+
+    @DaoFactory
+    SquareDao squareDao(@DaoKeyspace CqlIdentifier keyspace);
+
+    @DaoFactory
+    SphereDao sphereDao(@DaoKeyspace CqlIdentifier keyspace);
+
+    @DaoFactory
+    DeviceDao deviceDao(@DaoKeyspace CqlIdentifier keyspace, @DaoTable CqlIdentifier table);
+
+    @DaoFactory
+    TrackedDeviceDao trackedDeviceDao(
+        @DaoKeyspace CqlIdentifier keyspace, @DaoTable CqlIdentifier table);
+
+    @DaoFactory
+    SimpleDeviceDao simpleDeviceDao(@DaoKeyspace CqlIdentifier keyspace);
+  }
+
+  @DataProvider
+  public static Object[][] setAndGetProvider() {
+    Function<CqlIdentifier, CircleDao> circleDao = keyspace -> mapper.circleDao(keyspace);
+    Function<CqlIdentifier, RectangleDao> rectangleDao = keyspace -> mapper.rectangleDao(keyspace);
+    Function<CqlIdentifier, SquareDao> squareDao = keyspace -> mapper.squareDao(keyspace);
+    Function<CqlIdentifier, SphereDao> sphereDao = keyspace -> mapper.sphereDao(keyspace);
+    return new Object[][] {
+      new Object[] {
```

Comment:
```suggestion
      {
```
You don't need to repeat that for each inner array.